modules/sqs/main.tf

```diff
@@ -53,17 +53,81 @@ resource "aws_sqs_queue" "metadata_queue" {
-  message_retention_seconds   = var.staged_recovery_queue_message_retention_time_seconds
-  tags                        = local.tags
-  policy                      = data.aws_iam_policy_document.staged_recovery_queue_policy.json
-  visibility_timeout_seconds  = 1800 # Set to double lambda max time
+  name                       = "${var.prefix}-staged-recovery-queue"
+  delay_seconds              = var.sqs_delay_time_seconds
+  max_message_size           = var.sqs_maximum_message_size
+  message_retention_seconds  = var.staged_recovery_queue_message_retention_time_seconds
+  tags                       = local.tags
+  policy                     = data.aws_iam_policy_document.staged_recovery_queue_policy.json
+  visibility_timeout_seconds = 1800 # Set to double lambda max time
+  depends_on = [
+    aws_sqs_queue.staged_recovery_deadletter_queue
+  ]
+  redrive_policy = jsonencode({
+    deadLetterTargetArn = aws_sqs_queue.staged_recovery_deadletter_queue.arn,
+    maxReceiveCount     = 3 #if number of delayed messages in the queue is greater than this number, then the messages will be sent to its DLQ.
+  })
 }
 
+resource "aws_sqs_queue" "staged_recovery_deadletter_queue" {
+  name                       = "${var.prefix}-staged-recovery-deadletter-queue"
+  delay_seconds              = var.sqs_delay_time_seconds
+  max_message_size           = var.sqs_maximum_message_size
+  message_retention_seconds  = var.staged_recovery_queue_message_retention_time_seconds
+  tags                       = local.tags
+  visibility_timeout_seconds = 1800 # Set to same time as the staged recovery queue.
+}
+
+resource "aws_sqs_queue_policy" "deadletter_queue_policy" {
+  queue_url = aws_sqs_queue.staged_recovery_deadletter_queue.id
+  policy    = data.aws_iam_policy_document.staged_recovery_deadletter_queue_policy.json
+}
+
+data "aws_iam_policy_document" "staged_recovery_deadletter_queue_policy" {
+  statement {
+    effect    = "Allow"
+    resources = [aws_sqs_queue.staged_recovery_deadletter_queue.arn]
+    actions = [
+      "sqs:ChangeMessageVisibility",
+      "sqs:DeleteMessage",
+      "sqs:GetQueueAttributes",
+      "sqs:GetQueueUrl",
+      "sqs:ListQueueTags",
+      "sqs:ReceiveMessage",
+      "sqs:SendMessage",
+    ]
+  }
+}
+
+resource "aws_cloudwatch_metric_alarm" "staged_recovery_deadletter_alarm" {
+  alarm_name          = "${aws_sqs_queue.staged_recovery_deadletter_queue.name}-not-empty-alarm"
+  alarm_description   = "Items are on the ${aws_sqs_queue.staged_recovery_deadletter_queue.name} queue"
+  comparison_operator = "GreaterThanOrEqualToThreshold"
+  evaluation_periods  = 1
+  metric_name         = "ApproximateNumberOfMessagesVisible"
+  namespace           = "AWS/SQS"
+  period              = 300 #In seconds. Valid values for period are 1, 5, 10, 30, or any multiple of 60. 
+  statistic           = "Average"
+  threshold           = 1 #alarm will be triggered if number of messages in the DLQ equals to this threshold.
```

Comment:
Not strictly accurate. The statistic is "Average", not sum. So if a single message is in the queue and does not remain for the entire period, the threshold will not be tripped.

Reply:
it will send an alarm when threshold is 1. So actually average or sum will give same result.
see example
https://github.com/damacus/terraform-aws-sqs-with-dlq/blob/a041c4343d99a3c46a53dce8ac0fda2bc6559193/main.tf#L87

Reply:
I don't know what this link is intended to suggest. My understanding is that if one item is in the queue for 299 seconds, then the Average would be 1 * (299/300) which would not trigger the threshold.

Reply:
makes sense. I will have to test more to see what happens with sum.

If everyone agrees, I can change this to `sum`

Reply:
I think sum may be a better choice here.